packages/components/src/tooltip/test/index.tsx

```diff
@@ -0,0 +1,285 @@
+import userEvent from '@testing-library/user-event';
+
+/**
+ * WordPress dependencies
+ */
+import { shortcutAriaLabel } from '@wordpress/keycodes';
+
+/**
+ * Internal dependencies
+ */
+import Button from '../../button';
+import Modal from '../../modal';
+import Tooltip, { TOOLTIP_DELAY } from '..';
+import cleanupTooltip from './utils/';
+
+const props = {
+	children: <Button>Button</Button>,
+	text: 'tooltip text',
+};
+
+describe( 'Tooltip', () => {
+	it( 'should not render the tooltip if multiple children are passed', async () => {
+		render(
+			// expected TS error since Tooltip cannot have more than one child element
+			// @ts-expect-error
+			<Tooltip { ...props }>
+				<Button>This is a button</Button>
+				<Button>This is another button</Button>
+			</Tooltip>
+		);
+
+		expect(
+			screen.queryByRole( 'tooltip', { name: /tooltip text/i } )
+		).not.toBeInTheDocument();
+	} );
+
+	it( 'should not render the tooltip if there is no focus', () => {
+		render( <Tooltip { ...props } /> );
+
+		expect(
+			screen.getByRole( 'button', { name: /Button/i } )
+		).toBeVisible();
+
+		expect(
+			screen.queryByRole( 'tooltip', { name: /tooltip text/i } )
+		).not.toBeInTheDocument();
+	} );
+
+	it( 'should render the tooltip when focusing on the tooltip anchor via tab', async () => {
+		const user = userEvent.setup();
+
+		render( <Tooltip { ...props } /> );
+
+		await user.tab();
+
+		expect(
+			screen.getByRole( 'button', { name: /Button/i } )
+		).toHaveFocus();
+
+		expect(
+			await screen.findByRole( 'tooltip', { name: /tooltip text/i } )
+		).toBeVisible();
+
+		await cleanupTooltip( user );
+	} );
+
+	it( 'should render the tooltip when the tooltip anchor is hovered', async () => {
+		const user = userEvent.setup();
+
+		render( <Tooltip { ...props } /> );
+
+		await user.hover( screen.getByRole( 'button', { name: /Button/i } ) );
+
+		expect(
+			await screen.findByRole( 'tooltip', { name: /tooltip text/i } )
+		).toBeVisible();
+
+		await cleanupTooltip( user );
+	} );
+
+	it( 'should not show tooltip on focus as result of mouse click', async () => {
+		const user = userEvent.setup();
+
+		render( <Tooltip { ...props } /> );
+
+		await user.click( screen.getByRole( 'button', { name: /Button/i } ) );
+
+		expect(
+			screen.queryByRole( 'tooltip', { name: /tooltip text/i } )
+		).not.toBeInTheDocument();
+
+		await cleanupTooltip( user );
+	} );
+
+	it( 'should respect custom delay prop when showing tooltip', async () => {
+		const user = userEvent.setup();
+		const CUSTOM_DELAY = TOOLTIP_DELAY + 25;
+
+		render( <Tooltip { ...props } delay={ CUSTOM_DELAY } /> );
+
+		await user.hover( screen.getByRole( 'button', { name: /Button/i } ) );
+
+		expect(
+			screen.queryByRole( 'tooltip', { name: /tooltip text/i } )
+		).not.toBeInTheDocument();
+
+		expect(
+			await screen.findByRole( 'tooltip', { name: /tooltip text/i } )
+		).toBeVisible();
+
+		await cleanupTooltip( user );
+	} );
```

Comment:
I thought about this too, so I'm glad you brought it up! I thought it was testing that the custom delay passed, but I'm not sure how valuable it is just like that. 🤔 

We have a very similar test to make sure the tooltip doesn't show if the anchor leaves before the delay time, so along with your suggestions, we can also borrow from that: https://github.com/WordPress/gutenberg/blob/50b7f5dc20dc0393a04ffe10a21e5c48d512a652/packages/components/src/tooltip/test/index.tsx#L141

I'm not sure if it's better to wrap the promise in `waitFor` rather than `act`, but we can do that to avoid adding another method, and `waitFor` is already part of other tests. 

https://github.com/WordPress/gutenberg/blob/ed4c6aa3ae689cf15343d5c3f2cd19f7acc3669e/packages/components/src/tooltip/test/index.tsx#L119-L125

Reply:
Is the `waitFor` wrap necessary? In other instances in the same file, we've simply `await`ed for the promise, without `waitFor` or `act`

Reply:
If we use just `await` without `waitFor` in this case, we get `act()` warnings: 

>    `When testing, code that causes React state updates should be wrapped into act(...):`

But this is a bit of an odd test, and I'm still unsure of its value. When first looking at this test, it makes sense to incrementally advance the time to ensure it reaches the custom delay and doesn't display the tooltip before that. 

But if we set a custom delay on tooltip, expect to find the tooltip, and it passes, it seems that would be enough without manually setting timeouts. 

We could remove the second promise and use `findBy` which passes: 

```diff
--- a/packages/components/src/tooltip/test/index.tsx
+++ b/packages/components/src/tooltip/test/index.tsx
@@ -116,16 +116,8 @@ describe( 'Tooltip', () => {
 			screen.queryByRole( 'tooltip', { name: /tooltip text/i } )
 		).not.toBeInTheDocument();
 
-		// wait for additional delay for tooltip to appear
-		await waitFor(
-			() =>
-				new Promise( ( resolve ) =>
-					setTimeout( resolve, ADDITIONAL_DELAY )
-				)
-		);
-
 		expect(
-			screen.getByRole( 'tooltip', { name: /tooltip text/i } )
+			await screen.findByRole( 'tooltip', { name: /tooltip text/i } )
 		).toBeVisible();
 
 		await cleanupTooltip( user );
```

But we could also remove both promises, achieve the same result, and return to where it started. So I wonder: 

1. Is it valuable as it is with timeouts?
2. As it was without them?
3. Or do we need to find another way to test it? Like maybe checking if the time passed when tooltip appears is the same as the custom delay?

Reply:
Let's not delay this PR further and keep the test as-is now, we can always make improvements in a follow-up PR.

Cc'ing @jsnajdr if/when he has a moment to take a look at this, since he's already worked on fixing a few other aspects of tooltip unit tests

Reply:
If we expect React update inside a timeout, it must be wrapped in `act()`. Like:
```js
render( <Tooltip delay={ TOOLTIP_DELAY } /> );
await user.hover( screen.getByRole( 'button', { name: /Button/i } ) );

// an instant before timeout
await sleep( TOOLTIP_DELAY - EPSILON );
expect( screen.queryByRole( 'tooltip' ) ) .not.toBeInTheDocument();

// an instant after timeout
await act( () => sleep( 2 * EPSILON ) );
expect( screen.getByRole( 'tooltip' ) ).toBeInTheDocument();
```
This test is passing for me without any warnings.

It's also a good idea to use fake timers here and advance them manually. Makes the test more reliable and slightly faster, as it doesn't need to wait the real 700ms:
```js
render( <Tooltip delay={ TOOLTIP_DELAY } /> );
await user.hover( screen.getByRole( 'button', { name: /Button/i } ) );

jest.advanceTimersByTime( TOOLTIP_DELAY - EPSILON );
expect( screen.queryByRole( 'tooltip' ) ) .not.toBeInTheDocument();

await act( async () => jest.advanceTimersByTime( 2 * EPSILON ) );
expect( screen.getByRole( 'tooltip' ) ).toBeInTheDocument();
```
But when I tried this fake timers approach I ran into big trouble, because an `act` warning suddenly appears. It's caused by Ariakit's internal usage of Floating UI. There is a `setState` call similar to this:
```js
useEffect( () => {
  setPositioned( true );
}
```
Which sets state that's already `true` to `true`. Therefore it never has any observable effect, but the `act` warning is triggered by the `setState` being merely scheduled, no matter what is the result. I don't know how to fix it except trying random variants of `await act( () => waitForMicrotaskOrTimeoutOrAnimFrame() )`.

It's this line in Ariakit: https://github.com/ariakit/ariakit/blob/main/packages/ariakit-react-core/src/popover/popover.tsx#L294

Floating UI's native `update` function is very careful to not set the state unless it's sure that it's different:

https://github.com/floating-ui/floating-ui/blob/master/packages/react-dom/src/useFloating.ts#L101-L107

It could help if Ariakit started doing this in its `updatePosition2`, too. Cc @diegohaz